File: android/src/main/kotlin/com/ccpp/pgw_sdk/api/CardTokenInfoApi.kt
/*
 * Created by Nonthawatt Phongwittayapanu on 10/20/21, 11:31 AM
 * Copyright © 2021 2C2P. All rights reserved.
 */

package com.ccpp.pgw_sdk.api

import android.util.Log
import androidx.annotation.NonNull
import com.ccpp.pgw.sdk.android.callback.APIResponseCallback
import com.ccpp.pgw.sdk.android.core.PGWSDK
import com.ccpp.pgw.sdk.android.enums.APIResponseCode
import com.ccpp.pgw.sdk.android.model.api.CardTokenInfoRequest
import com.ccpp.pgw.sdk.android.model.api.CardTokenInfoResponse
import com.ccpp.pgw_sdk.helper.JsonHelper.jsonString
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class CardTokenInfoApi(
    @NonNull val call: MethodCall, @NonNull val result: MethodChannel.Result
) {

    /**
     * Card token info function
     */
    fun request() {
        val json = call.argument<String>("cardTokenInfoRequest") ?: ""
        val request = cardTokenInfoRequestFromJson(json) ?: kotlin.run {
            result.error("998", "Parse json error", null)
            return
        }

        PGWSDK.getInstance().cardTokenInfo(request, object : APIResponseCallback<CardTokenInfoResponse> {
            override fun onResponse(response: CardTokenInfoResponse) {
                if (response.responseCode == APIResponseCode.APISuccess) {
                    //Read payment option response.
                    result.success(cardTokenInfoResponseToJson(response))
                } else {
                    //Get error response and display error.
                    Log.d("DEMO", "onResponse: Error ${response.responseCode}")
                    Log.d("DEMO", "onResponse: Error ${response.responseDescription}")

                    result.error(response.responseCode, response.responseDescription, null)
                }
            }

            override fun onFailure(error: Throwable) {
                //Get error response and display error.
                Log.d("DEMO", "onFailure: $error")
                result.error("999", error.localizedMessage, error)
            }
        })
    }

    private fun cardTokenInfoRequestFromJson(json: String): CardTokenInfoRequest? {
        val jsonObject = JSONObject(json)

        Log.d("DEMO", "parseJson: Request =>> ${jsonObject.toString(2)}")

        val paymentToken = jsonString(jsonObject, "paymentToken")
        val clientId = jsonString(jsonObject, "clientID")
        val locale = jsonString(jsonObject, "locale")

        return CardTokenInfoRequest(paymentToken).apply {
            if (!clientId.isNullOrBlank()) {
                setClientId(clientId)
            }
            if (!locale.isNullOrBlank()) {
                setLocale(locale)
            }
        }
    }

    private fun cardTokenInfoResponseToJson(response: CardTokenInfoResponse): String {
        val cardTokens: ArrayList<Map<String, Any?>> = arrayListOf()
        response.cardTokens.forEach {
            cardTokens.add(
                mapOf(
                    "token" to it.token,
                    "cardNo" to it.cardNo,
                    "expiryDate" to it.expiryDate,
                    "name" to it.name,
                    "email" to it.email,
                    "status" to it.status,
                    "iconUrl" to it.iconUrl,
                    "logoUrl" to it.logoUrl,
                )
            )
        }

        val responseMap = mapOf(
            "paymentToken" to response.paymentToken,
            "cardTokens" to cardTokens,
            "responseCode" to response.responseCode,
            "responseDescription" to response.responseDescription,
        )

        return JSONObject(responseMap).toString()
    }

}

File: ios/Classes/Api/CardTokenInfoApi.swift
//
//  CardTokenInfoApi.swift
//  pgw_sdk
//
//  Created by Nonthawatt Phongwittayapanu on 20/10/2564 BE.
//

import UIKit
import PGW

class CardTokenInfoApi {
    private var call: FlutterMethodCall
    private var result: FlutterResult
    
    init(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        self.call = call
        self.result = result
    }
    
    func request() {
        guard let dicts = call.arguments as? [String: Any],
              let jsonString = dicts["cardTokenInfoRequest"] as? String
        else {
            result(FlutterError.init(code: "998", message: "Arguments exception", details: nil))
            return
        }
        guard let request = mappingRequest(json: jsonString) else {
            result(FlutterError.init(code: "900", message: "Json exception", details: nil))
            return
        }
        
        PGWSDK.shared.cardTokenInfo(cardTokenInfoRequest: request, { (response: CardTokenInfoResponse) in
            if response.responseCode == APIResponseCode.APISuccess {
                if let jsonString = self.mappingResposne(response) {
                    self.result(jsonString)
                } else {
                    self.result(FlutterError.init(code: response.responseCode, message: "Parse json error, \(response.responseDescription ?? "")", details: nil))
                }
            } else {
                //Get error response and display error.
                self.result(FlutterError.init(code: response.responseCode, message: response.responseDescription, details: nil))
            }
        }) { (error: NSError) in
            //Get error response and display error.
            self.result(FlutterError.init(code: "\(error.code)", message: error.localizedDescription, details: nil))
        }
    }
    
    // MARK: - Request
    private func mappingRequest(json: String) -> CardTokenInfoRequest? {
        do {
            let jsonData = json.data(using: .utf8)!
            let jsonDict = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any?]
            let paymentToken = jsonDict?["paymentToken"] as? String ?? ""
            
            let request = CardTokenInfoRequest(paymentToken: paymentToken)
            
            if let clientId = jsonDict?["clientID"] as? String {
                request.clientId = clientId
            }
            if let locale = jsonDict?["locale"] as? String {
                request.locale = locale
            }
            return request
        } catch let error as NSError {
            print(error)
            return nil
        }
    }
    
    // MARK: - Response
    private func mappingResposne(_ response: CardTokenInfoResponse) -> String? {
        
        var cardTokens: [[String: Any?]] = []
        response.cardTokens.forEach { (it: CardTokenInfo) in
            cardTokens.append(
                [
                    "token": it.token,
                    "cardNo": it.cardNo,
                    "expiryDate": it.expiryDate,
                    "name": it.name,
                    "email": it.email,
                    "status": it.status,
                    "iconUrl": it.iconUrl,
                    "logoUrl": it.logoUrl,
                ]
            )
        }
        
        let responseDict: [String : Any?] = [
            "paymentToken": response.paymentToken,
            "cardTokens": cardTokens,
            "responseCode": response.responseCode as String,
            "responseDescription": response.responseDescription as String,
        ]
        
        do {
            let data = try JSONSerialization.data(withJSONObject: responseDict)
            let string = String(data: data, encoding: .utf8)
            print(string ?? "")
            return string
        } catch let error as NSError {
            print(error)
            return nil
        }
    }
}
